app/services/school_data_importer.rb

```diff
@@ -46,77 +41,120 @@ def school_data_files
-    row_lad_name = row.fetch("DistrictAdministrative (name)")
+    def record_school_details
+      school.name = data.fetch("EstablishmentName")
+      school.address_line1 = data.fetch("Street")
+      school.address_line2 = data.fetch("Locality")
+      school.address_line3 = data.fetch("Town")
+      school.postcode = data.fetch("Postcode")
+      school.school_phase_type = data.fetch("PhaseOfEducation (code)").to_i
+      school.school_phase_name = data.fetch("PhaseOfEducation (name)")
+      school.school_website = data.fetch("SchoolWebsite")
+      school.primary_contact_email = data.fetch("MainEmail").presence
+      school.secondary_contact_email = data.fetch("AlternativeEmail").presence
 
-    if local_authority_district.persisted? && local_authority_district.name != row_lad_name
-      logger.info "LA name change in school import. Old name: #{local_authority_district.name}, New name: #{row_lad_name}"
+      # Changes to properties below carries major consequences and must be avoided until we know how to handle them
+      if new_school?
+        school.ukprn = data.fetch("UKPRN")
+
+        school.school_type_code = data.fetch("TypeOfEstablishment (code)").to_i
+        school.school_type_name = data.fetch("TypeOfEstablishment (name)")
+
+        school.school_status_code = data.fetch("EstablishmentStatus (code)").to_i
+        school.school_status_name = data.fetch("EstablishmentStatus (name)")
+
+        school.administrative_district_code = data.fetch("DistrictAdministrative (code)")
+        school.administrative_district_name = data.fetch("DistrictAdministrative (name)")
+      end
+
+      school.save!
     end
 
-    local_authority_district.name = row_lad_name
-    local_authority_district.save!
-    local_authority_district
-  end
+    def local_authority
+      return @local_authority if @local_authority
+
+      code = data.fetch("LA (code)")
+      @la_cache.fetch(code) do
+        local_authority = LocalAuthority.find_or_initialize_by(code: code)
+        new_name = data.fetch("LA (name)")
 
-  def link_school_to_local_authority(school, local_authority)
-    school_local_authority = SchoolLocalAuthority.find_or_initialize_by(school: school, local_authority: local_authority)
+        if local_authority.name != new_name
+          if local_authority.persisted?
+            logger.info "LA name changed in school import. Old name: #{local_authority.name}, New name: #{new_name}"
+          end
+          local_authority.name = new_name
+          local_authority.save!
+        end
 
-    if school_local_authority.new_record?
-      school_local_authority.start_year = start_year
-      SchoolLocalAuthority.latest.find_by(school: school)&.update!(end_year: start_year)
+        @local_authority = @la_cache[code] = local_authority
+      end
     end
 
-    school_local_authority.save!
-  end
+    def local_district_authority
+      return @local_district_authority if @local_district_authority
```

Comment:
Under what circumstances would this be set?

Reply:
If we call this twice. In the current code probably never, but it feels like something that should be memoized - without memoization this method (in my head) changes from a more complicated getter (noun) into a processing object (verb) which woudl probably require a name change (like `find_local_district_authority` or sth). It is purely opinionated though and I cannot really justify it by anything else than semantic.